File: src/evaluate.cpp
#include "methods/scan_context.h"
#include "methods/dbow.h"
#include "evaluator.h"

#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <message_filters/time_synchronizer.h>

#include <rosbag/bag.h>
#include <rosbag/view.h>

void getMethods(std::vector<BaseMethod *> &methods, ros::NodeHandle &nh)
{
    std::string method_name;

    if (nh.getParam("method", method_name))
    {
        double threshold, leaf_size;
        nh.getParam("threshold", threshold);
        if (method_name == "dbow")
            methods.push_back(new DBoW(threshold));
        else if (method_name == "context")
        {
            nh.getParam("leaf_size", leaf_size);
            methods.push_back(new ScanContext(threshold, leaf_size));
        }
        else
        {
            ROS_ERROR("Invalid method parameter. Use 'dbow' or 'context'.");
            exit(-1);
        }
    }
    else
    {
        // BaseMethod *method1 = new DBoW(2.2);
        // methods.push_back(method1);

        // BaseMethod *method2 = new ScanContext(0.06, 0.2);
        // methods.push_back(method2);

        // Find best threshold for Scan Context
        for (double threshold = 0.01; threshold < 0.1; threshold += 0.01)
        {
            BaseMethod *method = new ScanContext(threshold, 0.2);
            methods.push_back(method);
        }

        // Find best threshold for DBoW
        // for (double threshold = 1.8; threshold <= 2.5; threshold += 0.1)
        // {
        //     BaseMethod *method = new DBoW(threshold);
        //     methods.push_back(method);
        // }
    }
}

void processBag(Evaluator &evaluator, std::string input_bag_path, std::string lidar_topic, std::string odom_topic, std::string camera_topic)
{
    rosbag::Bag bag;
    try
    {
        bag.open(input_bag_path, rosbag::bagmode::Read);
    }
    catch (rosbag::BagException &e)
    {
        ROS_ERROR("Error opening bag file: %s", e.what());
    }

    // Specify topics to read from the rosbag
    std::vector<std::string> topics = {lidar_topic, odom_topic, camera_topic};
    rosbag::View view(bag, rosbag::TopicQuery(topics));

    sensor_msgs::PointCloud2::ConstPtr cloud_msg;
    nav_msgs::Odometry::ConstPtr odom_msg;
    sensor_msgs::Image::ConstPtr img_msg;

    for (const rosbag::MessageInstance &m : view)
    {
        if (m.getTopic() == lidar_topic)
        {
            cloud_msg = m.instantiate<sensor_msgs::PointCloud2>();
        }
        else if (m.getTopic() == odom_topic)
        {
            odom_msg = m.instantiate<nav_msgs::Odometry>();
        }
        else if (m.getTopic() == camera_topic)
        {
            img_msg = m.instantiate<sensor_msgs::Image>();
        }

        if (cloud_msg && odom_msg && img_msg)
        {
            evaluator.synced_callback(cloud_msg, odom_msg, img_msg);
            cloud_msg.reset();
            odom_msg.reset();
            img_msg.reset();
        }
    }

    bag.close();
}

int main(int argc, char **argv)
{
    // Initialize ROS node
    ros::init(argc, argv, "evaluation_node");
    ros::NodeHandle nh;

    // Retrieve parameters from the ROS parameter server
    std::string method_name;
    int record_size, keyframe_type, cloud_size, min_idx_diff, max_angle_deg;
    double max_dist, threshold;
    bool angle_consideration, save_candidates;

    // Merge params
    nh.getParam("keyframe_type", keyframe_type);
    nh.getParam("cloud_size", cloud_size);

    // Common params
    nh.getParam("record_size", record_size);
    nh.getParam("min_idx_diff", min_idx_diff);
    nh.getParam("angle_consideration", angle_consideration);
    nh.getParam("max_angle_deg", max_angle_deg);
    nh.getParam("max_dist", max_dist);
    nh.getParam("save_candidates", save_candidates);

    if (keyframe_type == 1)
        record_size /= cloud_size;

    // Get the method parameters
    std::vector<BaseMethod *> methods;
    getMethods(methods, nh);

    // Create the evaluation object based on the selected methods
    Evaluator evaluator(methods, record_size, angle_consideration, max_angle_deg, max_dist, min_idx_diff, save_candidates);

    // Get merged topic names
    std::string lidar_topic, odom_topic, camera_topic;
    nh.getParam("merged_lidar_topic", lidar_topic);
    nh.getParam("merged_odom_topic", odom_topic);
    nh.getParam("merged_camera_topic", camera_topic);

    // Get rosbag params
    bool use_rosbag;
    std::string input_bag_path;
    nh.getParam("use_rosbag", use_rosbag);
    nh.getParam("merge_rosbag_output_path", input_bag_path);

    if (use_rosbag)
    {
        processBag(evaluator, input_bag_path, lidar_topic, odom_topic, camera_topic);
    }
    else
    {
        message_filters::Subscriber<sensor_msgs::PointCloud2> cloud_sub(nh, lidar_topic, 20);
        message_filters::Subscriber<nav_msgs::Odometry> odom_sub(nh, odom_topic, 100);
        message_filters::Subscriber<sensor_msgs::Image> img_sub(nh, camera_topic, 20);

        // Synchronize messages from different topics
        message_filters::TimeSynchronizer<sensor_msgs::PointCloud2, nav_msgs::Odometry, sensor_msgs::Image> sync(cloud_sub, odom_sub, img_sub, 100);

        // Register the synchronized callback
        sync.registerCallback(boost::bind(&Evaluator::synced_callback, &evaluator, _1, _2, _3));

        // Process callbacks and wait for new messages
        ros::spin();
    }

    return 0;
}

File: scripts/evaluate.py
#!/usr/bin/env python

import rospy
import rosbag
from sensor_msgs.msg import PointCloud2, Image
from nav_msgs.msg import Odometry
from message_filters import TimeSynchronizer, Subscriber

from scripts.evaluator import Evaluator
from typing import List
from scripts.methods.base import BaseMethod

def getMethods() -> List[BaseMethod]:
    """
    Get the methods to be evaluated.

    :return: List[BaseMethod], the list of methods to be evaluated
    """
    # Create methods
    methods = []
    method_name = rospy.get_param('method', None)

    if method_name is not None:
        threshold = float(rospy.get_param('threshold'))

        if method_name == "logg3d":
            from scripts.methods.LoGG3D import LoGG3D
            methods.append(LoGG3D(threshold))
        elif method_name == "superglue":
            from scripts.methods.SuperGlue import SuperGlue
            methods.append(SuperGlue(threshold))
        elif method_name == "mix_vpr":
            from scripts.methods.MixVPR_ import MixVPR
            methods.append(MixVPR(threshold))
        else:
            rospy.logerr(
                "Invalid method parameter. Use 'logg3d' or 'superglue'.")
            exit(-1)
    else:
        # from scripts.methods.LoGG3D import LoGG3D
        # methods.append(LoGG3D(0.08))

        # from scripts.methods.SuperGlue import SuperGlue
        # methods.append(SuperGlue(240))

        # from scripts.methods.MixVPR_ import MixVPR
        # methods.append(MixVPR(0.13))

        # Example for finding best threshold for LoGG3D
        from scripts.methods.LoGG3D import LoGG3D
        for threshold in range(5, 17):
            threshold = threshold / 100.0  # Convert to float equivalent
            methods.append(LoGG3D(threshold))

        # Example for finding best threshold for SuperGlue
        # from scripts.methods.SuperGlue import SuperGlue
        # for threshold in range(150, 301, 30):
        #     methods.append(SuperGlue(threshold))

        # Example for finding best threshold for MixVPR
        # from scripts.methods.MixVPR_ import MixVPR
        # for threshold in range(9, 21):
        #     threshold = threshold / 100.0  # Convert to float equivalent
        #     methods.append(MixVPR(threshold))

    return methods


def process_bag(evaluator, input_bag_path, lidar_topic, odom_topic, camera_topic):
    cloud_msg = None
    odom_msg = None
    img_msg = None

    with rosbag.Bag(input_bag_path, 'r') as bag:
        for topic, msg, t in bag.read_messages(topics=[lidar_topic, odom_topic, camera_topic]):
            if topic == lidar_topic:
                cloud_msg = msg
            elif topic == odom_topic:
                odom_msg = msg
            elif topic == camera_topic:
                img_msg = msg

            if cloud_msg is not None and odom_msg is not None and img_msg is not None:
                evaluator.synced_callback(cloud_msg, odom_msg, img_msg)
                cloud_msg = None
                odom_msg = None
                img_msg = None


rospy.init_node('evaluation_node')

# Retrieve parameters from the ROS parameter server
keyframe_type = int(rospy.get_param('keyframe_type'))
cloud_size = int(rospy.get_param('cloud_size'))

# Common params
record_size = int(rospy.get_param('record_size'))
min_idx_diff = int(rospy.get_param('min_idx_diff'))
angle_consideration = bool(rospy.get_param('angle_consideration'))
max_angle_deg = int(rospy.get_param('max_angle_deg'))
max_dist = float(rospy.get_param('max_dist'))
save_candidates = bool(rospy.get_param('save_candidates'))

if keyframe_type == 1:
    record_size = int(record_size / cloud_size)

# Get methods
methods = getMethods()

# Create evaluator
evaluator = Evaluator(methods, record_size, angle_consideration,
                      max_angle_deg, max_dist, min_idx_diff, save_candidates)


# Get merged topics
lidar_topic = rospy.get_param('merged_lidar_topic')
odom_topic = rospy.get_param('merged_odom_topic')
camera_topic = rospy.get_param('merged_camera_topic')

use_rosbag = bool(rospy.get_param('use_rosbag'))
input_bag_path = rospy.get_param('merge_rosbag_output_path')

if use_rosbag:
    process_bag(evaluator, input_bag_path,
                lidar_topic, odom_topic, camera_topic)
else:
    # Create subscribers
    cloud_sub = Subscriber(lidar_topic, PointCloud2, queue_size=5)
    odom_sub = Subscriber(odom_topic, Odometry, queue_size=20)
    img_sub = Subscriber(camera_topic, Image, queue_size=5)

    # Synchronize messages from different topics
    sync = TimeSynchronizer([cloud_sub, odom_sub, img_sub], queue_size=10)

    # Register the synchronized callback
    sync.registerCallback(evaluator.synced_callback)

    # Process callbacks and wait for new messages
    rospy.spin()
